File: L/2958-length-of-longest-subarray-with-at-most-k-frequency/main.go
func maxSubarrayLength(nums []int, k int) int {
	m := map[int][]int{}
	ans := 0
	start := -1
	for i, n := range nums {
		m[n] = append(m[n], i)
		if len(m[n]) > k {
			if ans < i-1-start {
				ans = i - 1 - start
			}
			if start < m[n][len(m[n])-1-k] {
				start = m[n][len(m[n])-1-k]
			}
		}
	}
	if ans < len(nums)-start-1 {
		ans = len(nums) - start - 1
	}
	return ans
}

File: L/2958-length-of-longest-subarray-with-at-most-k-frequency/main.rs
impl Solution {
  pub fn max_subarray_length(nums: Vec<i32>, k: i32) -> i32 {
    let mut m: std::collections::HashMap<i32, i32> =
      std::collections::HashMap::with_capacity(nums.len());
    let mut start: usize = 0;
    let mut ans: usize = 0;
    for (idx, &v) in nums.iter().enumerate() {
      *m.entry(v).or_insert(0) += 1;

      while *m.get(&v).unwrap() > k {
        *m.get_mut(&nums[start]).unwrap() -= 1;
        start += 1;
      }
      ans = ans.max(idx - start + 1);
    }

    ans as i32
  }
}

struct Solution {}
